refactor(localize): update version of Babel

JoostK opened the pull request:
The `@angular/localize` package depends on a version of Babel that is two years
old, so this commit updates to the latest version.

Some changes were made to the linker and compliance tests to account for slight
changes in source maps, along with a few code updates because of changes to
the typings of Babel.

JoostK committed:
refactor(localize): update version of Babel

The `@angular/localize` package depends on a version of Babel that is two years
old, so this commit updates to the latest version.

Some changes were made to the linker and compliance tests to account for slight
changes in source maps, along with a few code updates because of changes to
the typings of Babel.

JoostK commented on packages/compiler-cli/linker/babel/src/babel_declaration_scope.ts:
```diff
@@ -11,7 +11,8 @@ import {DeclarationScope} from '../../../linker';
 
-export type ConstantScopePath = NodePath<t.Function|t.Program>;
+export type ConstantScopePath =
+    NodePath<t.FunctionDeclaration>|NodePath<t.FunctionExpression>|NodePath<t.Program>;
```
Splitting the type to become a union allows for better type-narrowing, e.g. after `path.isProgram()` has been used TS can now narrow to `NodePath<t.FunctionDeclaration>|NodePath<t.FunctionExpression>`.

JoostK commented on packages/compiler-cli/linker/babel/src/babel_declaration_scope.ts:
```diff
@@ -56,10 +57,11 @@ export class BabelDeclarationScope implements DeclarationScope<ConstantScopePath
     // `t.Program` that is marked as a module).
     const path = binding.scope.path;
-    if (!path.isFunctionParent() && !(path.isProgram() && path.node.sourceType === 'module')) {
+    if (!path.isFunctionDeclaration() && !path.isFunctionExpression() &&
+        !(path.isProgram() && path.node.sourceType === 'module')) {
```
This was changed as `isFunctionParent` also includes functions that don't necessarily have a block statement as body, which means that we wouldn't be able to insert extra statements into it.

JoostK commented on packages/compiler-cli/test/compliance/test_helpers/expect_emit.ts:
```diff
@@ -13,7 +13,7 @@ const COMMENT_END = /\*\//;
     /!|\?|%|\*|\/|\^|&&?|\|\|?|\(|\)|\{|\}|\[|\]|:|;|<=?|>=?|={1,3}|!==?|=>|\+\+?|--?|@|,|\.|\.\.\.|\\`|\\'/;
 const STRING = /'(\\'|[^'])*'|"(\\"|[^"])*"/;
-const INLINE_BACKTICK_STRING = /`(([\s\S]*?)(\$\{[^}]*?\})?)*?[^\\]`/;
+const INLINE_BACKTICK_STRING = /`(?:[\s\S]|(?:\$\{[^}]*?\}))*?[^\\]`/;
```
This was updated because the original regex would never terminate when matched in V8 if the closing backtick is missing.

alan-agius4 commented:
Once this is green you should likely also remove `@babel/` packages from https://github.com/angular/angular/blob/cae1e44608227ac80682401190c6d28ef39b3269/renovate.json#L14

JoostK committed:
ci: allow Babel to be updated by Renovate

Automatic updating of Babel was disabled because there were build failures
when using the latest version of Babel, but these have been resolved in
#44931. This commit removes Babel packages from Renovate's ignored list
to allow automatic updates again.

JoostK commented on packages/localize/package.json:
```diff
@@ -34,7 +34,7 @@
   ],
   "dependencies": {
-    "@babel/core": "7.8.6",
+    "@babel/core": "7.16.12",
```
Should we consider unpinning this dep? The linker has an unpinned dep already.

alan-agius4 replied:
I’d suggest to keep for the extract version that is used in the root package.json so that it depends on the same version that we test with.

JoostK replied:
Okay, then WDYT about pinning the linker's Babel dep?

alan-agius4 replied:
I am good with that.

AndrewKushnir reviewed (approved):
👍

devversion commented on packages/compiler-cli/private/babel.d.ts:
```diff
@@ -0,0 +1,42 @@
+import { BabelFile, types as t } from '@babel/core';
+
+/**
+ * Augment some Babel types to add symbols that we rely on, but are not included in the Babel typings.
```
Is there an issue for this somewhere? (something we could reference with a todo?)

JoostK replied:
I have been trying to setup a dedicated repro to report in the rules_nodejs repo, but I'm still experiencing the absolute weirdest and inconsistent results. It repros _sometimes_, and `bazel clean` doesn't seem to consistently make a difference (nor does `shutdown`/`clean --expunge`). Even more bizarre, running the command reported when using the `-s` flag _does_ fail with the expected error, so I'm absolutely confused as to why Bazel itself succeeds (running Bazel with `--spawn_strategy=local`)

JoostK replied:
i.e. this makes no sense to me 🤯 

<img width="1626" alt="image" src="https://user-images.githubusercontent.com/123679/154350851-d9046168-dedf-4d75-b838-286ed0bdd92d.png">

JoostK replied:
Here's the repro if you're interested: https://github.com/JoostK/bazel_strict_deps

devversion commented on packages/compiler-cli/private/babel.d.ts:
```diff
@@ -0,0 +1,42 @@
+// `node_modules` and therefore the `@types` end up being resolved from a `node_modules` tree
+// that is not governed by Bazel, and therefore not expected by the `strict_deps` rule.
+// Declaring the modules here allows `strict_deps` to always find a declaration of the modules
+// in an input file to the compilation, therefore accepting the module import.
```
Awesome comment! this thing is still painful to see though 😄 wish we could address this in the future.

devversion reviewed (approved):
LGTM for the dev-infra related stuff

jessicajaniuk commented:
@JoostK Looks like google3 has some big issues with this PR.

JoostK commented:
> @JoostK Looks like google3 has some big issues with this PR.

I had big issues getting this green on CI as well 🙃

I thought Babel wasn't being used in g3 and I'd be interested to learn how this fails.

jessicajaniuk commented:
Alas, this PR does not merge cleanly with `13.2.x`. I'll update to `target: minor` and land it. Please create a patch PR, if you'd like to merge into `13.2.x`.

jessicajaniuk commented:
This PR was merged into the repository by commit 0debf840456cdc9f6bb3976ab5b50e194d5127f5.
